src/index/flat/flat.cc

```diff
@@ -111,13 +108,13 @@ class FlatIndexNode : public IndexNode {
 
                         index_->search(1, cur_query, k, cur_dis, cur_ids, &search_params);
                     }
-                    if constexpr (std::is_same<T, faiss::IndexBinaryFlat>::value) {
+                    if constexpr (std::is_same<IndexType, faiss::IndexBinaryFlat>::value) {
                         auto cur_i_dis = reinterpret_cast<int32_t*>(cur_dis);
 
                         faiss::SearchParameters search_params;
                         search_params.sel = id_selector;
 
-                        index_->search(1, (const uint8_t*)x + index * dim / 8, k, cur_i_dis, cur_ids, &search_params);
+                        index_->search(1, (const DataType*)x + index * dim / 8, k, cur_i_dis, cur_ids, &search_params);
```

Comment:
this is pretty confusing piece of code, because it literally implies that `DataType` is `uint8_t` because of the following `dim / 8`, and if `DataType` is `uint32_t`, then this code will fail.
Same note for this whole file.
Maybe, put an Index that uses `faiss::IndexBinaryFlat` into a separate class?

Reply:
updated